cpp/test/Ice/metrics/AllTests.cpp

```diff
@@ -1144,19 +954,29 @@ allTests(Test::TestHelper* helper, const CommunicatorObserverIPtr& obsv)
         test(false);
     }
 
-    metrics->opWithUserExceptionAsync(
-        [cb]()
-        {
-            test(false);
-        },
-        [cb](exception_ptr e)
-        {
-            cb->exception(e);
-        });
-    cb->waitForResponse();
+    {
+        std::promise<void> p;
+        metrics->opWithUserExceptionAsync(
```

Comment:
Why not just use the future API here?

Reply:
We test with the 3 APIs

```
  try
    {
        metrics->opWithUserException();
        test(false);
    }
    catch(const Test::UserEx&)
    {
    }

    try
    {
        metrics->opWithUserExceptionAsync().get();
        test(false);
    }
    catch(const Test::UserEx&)
    {
    }
    catch(...)
    {
        test(false);
    }

    {
        std::promise<void> p;
        metrics->opWithUserExceptionAsync(
            []() { test(false); },
            [&p](exception_ptr e)
            {
                try
                {
                    rethrow_exception(e);
                }
                catch(const Test::UserEx&)
                {
                    p.set_value();
                }
                catch(...)
                {
                    test(false);
                }
            });
        p.get_future().wait();
    }
```

Maybe not require, but I didn't change that.  We should probably cleanup this in a future PR!